substrate/client/network/sync/src/pending_responses.rs

```diff
@@ -0,0 +1,114 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [`PendingResponses`] is responsible for keeping track of pending responses and
+//! polling them.
+
+use futures::{
+	channel::oneshot,
+	future::BoxFuture,
+	stream::{BoxStream, Stream},
+	FutureExt, StreamExt,
+};
+use libp2p::PeerId;
+use log::error;
+use sc_network::request_responses::RequestFailure;
+use sc_network_common::sync::PeerRequest;
+use sp_runtime::traits::Block as BlockT;
+use std::task::{Context, Poll};
+use tokio_stream::StreamMap;
+
+/// Response result.
+type ResponseResult = Result<Result<Vec<u8>, RequestFailure>, oneshot::Canceled>;
+
+/// A future yielding [`ResponseResult`].
+type ResponseFuture = BoxFuture<'static, ResponseResult>;
+
+/// An event we receive once a pending response future resolves.
+pub(crate) struct ResponseEvent<B: BlockT> {
+	pub peer_id: PeerId,
+	pub request: PeerRequest<B>,
+	pub response: ResponseResult,
+}
+
+/// Stream taking care of polling pending responses.
+pub(crate) struct PendingResponses<B: BlockT> {
+	/// Pending responses
+	pending_responses: StreamMap<PeerId, BoxStream<'static, (PeerRequest<B>, ResponseResult)>>,
+}
+
+impl<B: BlockT> PendingResponses<B> {
+	pub fn new() -> Self {
+		Self { pending_responses: StreamMap::new() }
+	}
+
+	pub fn insert(
+		&mut self,
+		peer_id: PeerId,
+		request: PeerRequest<B>,
+		response_future: ResponseFuture,
+	) {
+		let request_type = request.get_type();
+
+		if self
+			.pending_responses
+			.insert(
+				peer_id,
+				Box::pin(async move { (request, response_future.await) }.into_stream()),
+			)
+			.is_some()
+		{
+			error!(
+				target: crate::LOG_TARGET,
+				"Discarded pending response from peer {peer_id}, request type: {request_type:?}.",
+			);
+			debug_assert!(false);
+		}
+	}
+
+	pub fn remove(&mut self, peer_id: &PeerId) -> bool {
+		self.pending_responses.remove(peer_id).is_some()
+	}
+
+	pub fn len(&self) -> usize {
+		self.pending_responses.len()
+	}
+}
+
+impl<B: BlockT> Unpin for PendingResponses<B> {}
+
+impl<B: BlockT> Stream for PendingResponses<B> {
+	type Item = ResponseEvent<B>;
+
+	fn poll_next(
+		mut self: std::pin::Pin<&mut Self>,
+		cx: &mut Context<'_>,
+	) -> Poll<Option<Self::Item>> {
+		match futures::ready!(self.pending_responses.poll_next_unpin(cx)) {
+			Some((peer_id, (request, response))) => {
+				// We need to manually remove the stream, because `StreamMap` doesn't know yet that
+				// it's going to yield `None`, so may not remove it before the next request is made
+				// to the same peer.
+				self.pending_responses.remove(&peer_id);
```

Comment:
I forgot that I enabled auto-merge, and the PR got merged yesterday. The problem with this `remove()` is that internally `StreamMap` keeps stream in a vec, and performs a linear search when removing by key:
https://github.com/tokio-rs/tokio/blob/master/tokio-stream/src/stream_map.rs#L470-L482

The same applies to `PendingResponses::remove` above. We may need to optimize this.

Reply:
Looking at the implementation of `StreamMap::insert()`, I see they call the same `remove()` on every insertion. So this is also not super optimal.

Reply:
I'm not very concerned about this. We need `StreamMap` kind of semantics from whatever the storage is that holds pending responses and since there's at most 40 pending responses, I'm not sure we should dedicate time to optimizing this now.

It can definitely be improved, I agree with you on that but this is a shortcoming of the request-response API by not providing a good API for polling multiple responses in a way that any individual response can be canceled. Let's fix the issue in that API rather than addressing it in a consumer of the API. I'll start working on the request-response code soon